plugins-builtin/src/main/scala/za/co/absa/enceladus/plugins/builtin/errorsender/mq/KafkaErrorSenderPluginImpl.scala

```diff
@@ -0,0 +1,178 @@
+import za.co.absa.enceladus.plugins.builtin.errorsender.params.ErrorSenderPluginParams
+import za.co.absa.enceladus.plugins.builtin.errorsender.params.ErrorSenderPluginParams.ErrorSourceId
+import za.co.absa.enceladus.utils.error.ErrorMessage.ErrorCodes
+
+import scala.util.{Failure, Success, Try}
+
+
+case class KafkaErrorSenderPluginImpl(connectionParams: KafkaConnectionParams,
+                                      keySchemaRegistryConfig: Map[String, String],
+                                      valueSchemaRegistryConfig: Map[String, String]) extends PostProcessor {
+
+  private val log = LogManager.getLogger(classOf[KafkaErrorSenderPluginImpl])
+
+  override def close(): Unit = {}
+
+  /**
+   * When data is ready, the error record(s) are pusblished to kafka.
+   *
+   * @param dataFrame error data only.
+   * @param params    Additional key/value parameters provided by Enceladus.
+   */
+  override def onDataReady(dataFrame: DataFrame, params: Map[String, String]): Unit = {
+    if (!SchemaUtils.fieldExists(ColumnNames.enceladusRecordId, dataFrame.schema)) {
+      throw new IllegalStateException(
+        s"${this.getClass.getName} requires ${ColumnNames.enceladusRecordId} column to be present in the dataframe!"
+      )
+    }
+
+    val errorSenderParams = Try(ErrorSenderPluginParams.fromMap(params)) match {
+      case Success(params) => params
+      case Failure(e) =>
+        throw new IllegalArgumentException(s"Incompatible parameter map supplied for ${KafkaErrorSenderPluginImpl.getClass.getName}: $params", e)
+    }
+
+    val dfWithErrors = getIndividualErrors(dataFrame, errorSenderParams)
+    val forKafkaDf = prepareDataForKafka(dfWithErrors, errorSenderParams)
+    sendErrorsToKafka(forKafkaDf)
+  }
+
+  /**
+   * Processes the `dataFrame` - errors are exploded (one error = one line) and filtered to conform to the error source (standardization/conformance)
+   * @param dataFrame standardization/conformance output
+   * @param params plugin processing parameters
+   * @return DF with exploded errors and corresponding to the given error source
+   */
+  def getIndividualErrors(dataFrame: DataFrame, params: ErrorSenderPluginParams): DataFrame = {
+    implicit val singleErrorStardardizedEncoder = Encoders.product[SingleErrorStardardized]
+    implicit val dceErrorEncoder = Encoders.product[DceError]
+
+    val allowedErrorCodes = KafkaErrorSenderPluginImpl.errorCodesForSource(params.sourceId)
+
+    val stdErrors = dataFrame
+      .filter(size(col("errCol")) > 0)
+      // only keep columns that are needed for the actual error publishing
+      .select(
+        col(ColumnNames.enceladusRecordId).cast(DataTypes.StringType).as("recordId"),
+        col(ColumnNames.reportDate),
+        explode(col(ColumnNames.errCol)).as("singleError")
+      )
+      .as[SingleErrorStardardized]
+      .map(_.toDceError(params))
+      .filter(entry => allowedErrorCodes.contains(entry.errorCode)) // Std xor Conf error codes
+      .toDF()
```

Comment:
Why convert the dataframe to a dataset (using the case class) if you are converting it back to a dataframe?
You pay for serialization/deserialization in this case, right?

To filter out error codes you can use pure Spark SQL DSL, something like:
```scala
val codesExpr = allowedErrorCodes.foldLeft(lit(false))((acc, code) =>
  acc || array_contains(col("errCol.errCode"), code)
)

df.where(codesExpr)
```

Reply:
The main reason to do this isn't the filtering by `errCode`, that indeed could be done as you suggest and it is not too much hassle. For me, the real reason is:
```scala
      .as[SingleErrorStardardized]
      .map(_.toDceError(params))
```
i.e. mapping non-trivial one data type to another, while having the clear benefit of compile-time safety as much as possible. I'd say that the same would probably be also possible with the DataFrame API (without having to convert to Dataset and back), however, it would surely be much uglier and less readable, not to mention less checked in compile time.

To be completely honest, I am not exactly sure what the cost of interchanging the Dataset/DataFrame is. I have been living under the impression that it is not acceptable with regard to the advantages gained.

I agree that just for the filtering, it might have been overkill, but to map between classes, it makes sense to me.

Maybe I need to lookup more a bit on the inner workings to get a more intimate notion of the costs of such conversion.

Reply:
Cool, thanks for such a detailed explanation. I like your solution.